keps/sig-instrumentation/4785-resource-state-metrics/README.md

```diff
@@ -0,0 +1,1216 @@
+  possible.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+N/A since the managed resource offered by Resource State Metrics provides the
+ability to define metric configurations that are a super-set of the 
+expressibility that Kube State Metrics' Custom Resource State configurations
+have to offer. However, users will need to familiarize themselves with the 
+supported DSLs to be able to translate their configuration into a format that
+the controller can understand, missing which could result in invalid metrics.
+
+Also, since this aims to deprecate the Custom Resource State API, a timeline 
+will be published for users on the Kube State Metrics' repository and a PSA on 
+the appropriate channel(s), once the controller graduates to stable.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The controller offers a number of improvements over Kube State Metrics' Custom
+Resource State API, while maintaining a 3x faster round trip time for metric
+generation.
+
+- At its core, the controller relies on its managed resource,
+`ResourceMetricsMonitor` to fetch the metric generation configuration. Parts of the
+configuration may be defined using different `resolver`s, such as `unstructured`
+or `CEL`.
+- Once fetched, the controller `unmarshal`s the configuration YAML directly into
+`stores` which are a set of metric `families`, which in turn are a set of 
+`metrics`.
+- Metric `stores` are created based on its respective GVKR (a type that embeds 
+`schema.GroupVersionKind`, `schema.GroupVersionResource` to avoid 
+[plural ambiguities]), and reflectors for the specified resource are
+initialized, and populate the stores on its update.
+- `/metrics` pings on `RSM_MAIN_PORT` trigger the server to write the
+raw metrics, combined with its appropriate header(s), in the response. All
+generated metrics are hardcoded to `gauge`s by design, as Prometheus lacks
+support for some OpenMetrics-specified metrics' types, such as `Info` and
+`StateSets`.
+
+At the moment, the `spec` houses a single `configuration` field, which defines
+the metric generation configuration as follows (please note that the schema is
+fast-moving at this point and may be subject to change based on the [feedback 
+obtained](https://github.com/rexagod/resource-state-metrics/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen)):
```

Comment:
Will it be considired to be changed in future and have the configuration be a typed api?

WIth having a `.spec.configuration` field which is a string, we could also just use a configmap or secret instead 🤔 

Also to build tooling on top (e.g. to generate a configuration for a CRD from markers) it would be helpful to have this an typed API.

Reply:
I like the idea of taking a config parameter reference - eg
```yaml
  configurationReference:
    apiVersion: v1       # for alpha, has to be ConfigMap
    kind: ConfigMap      # for alpha, has to be ConfigMap
    namespace: example   # optional, but must be set for initial / alpha implementation
    name: rsm-config
    itemPath: >-
      data.config
```

Reply:
Absolutely, I plan to make the `configuration` structured in the next CRD version, but  a reference to an object is doable right now, will do!